### PR TITLE
Light Switch Widget

### PR DESCRIPTION
**Pull request by mipolansk**
Light switch widget  implementation proposal. The design is not finalized yet, but would be good the have the logic reviewed.

### CODE REVIEW

File: app/src/main/java/org/supla/android/widget/SwitchPreferences.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+internal fun getKeyForChannelFunction(widgetId: Int): String {
+    return "$SHARED_PREFERENCES.CHANNEL_FUNCTION.$widgetId"
+}
+
+internal fun getKeyForChannelColor(widgetId: Int): String {
+    return "$SHARED_PREFERENCES.CHANNEL_COLOR.$widgetId"
+}
+
+data class WidgetConfiguration(
+        val channelId: Int,
```
Comment (przemyslawzygmunt) [inline app/src/main/java/org/supla/android/widget/SwitchPreferences.kt:82]: Here you will also need the profile id (profiles will be introduced soon)

File: app/src/main/java/org/supla/android/widget/lightswitch/LightSwitch.kt [hunk 2]
```diff
@@ -0,0 +1,155 @@
+package org.supla.android.widget.lightswitch
```
Comment (przemyslawzygmunt) [inline app/src/main/java/org/supla/android/widget/lightswitch/LightSwitch.kt:1]: I'm not sure if it should be "LightSwitch" or "OnOffSwitch". If OnOff, SUPLA_CHANNELFNC_POWERSWITCH can be added to the function list. The icon should match the default icon assigned to the channel function. User uploaded icons should be included.

File: app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+        val configuration = SwitchPreferences(applicationContext).getWidgetConfiguration(widgetIds[0])
+        if (configuration == null) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        showToast(R.string.widget_light_switch_command_started)
+
+        if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH) {
```
Comment (przemyslawzygmunt) [inline app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt:67]: With on/off for RGBW, it is enough to pass the brightness 100 for On and 0 for Off. The dimmer, if it sees the "OnOff" flag, will change 100 to the value memorized before switching it off.   

     if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH) {
            suplaClient.turnOnOff(applicationContext, inputData.getBoolean(ARG_TURN_ON, false),
                    configuration.channelId, false, configuration.channelFunction, false)
        } else if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_DIMMER
                || configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_DIMMERANDRGBLIGHTING
                || configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_RGBLIGHTING ) {
            suplaClient.setRGBW(configuration.channelId, configuration.channelColor, 100, 100, true)
        }

File: app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        if (suplaApp.suplaClient == null) {
+            suplaApp.SuplaClientInitIfNeed(applicationContext)
+        }
+        return suplaApp.suplaClient
+    }
+
+    private fun getBrightness(): Int {
+        return if (inputData.getBoolean(ARG_TURN_ON, false)) {
```
Comment (przemyslawzygmunt) [inline app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt:95]: It seems to me that the user question at onoff about the brightness level is unnecessary. However, if you want to ask about it, you should also ask about the color.

Comment (mipolansk) [reply app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt]: It's not user question. Widget has two buttons -> on and off. For on I'm setting brightness to 100 and for off to 0.

**Review by przemyslawzygmunt — APPROVED** [review]

**Pull request merged**